File: packages/models/src/normalize.ts
import type { Json, Message, Part } from "@goondan/core";
import { invalidRequest, unsupportedContent, type ModelProvider } from "./errors.ts";
import { isRecord } from "./json.ts";
import type { MediaResolver } from "./options.ts";

export type TextLikePart = Extract<Part, { type: "text" | "json" }>;
export type MediaPart = Extract<Part, { type: "media" }>;

/** Text of a `text` part, or the compact `JSON.stringify` text of a `json` part. */
export function partText(part: TextLikePart): string {
  return part.type === "text" ? part.text : JSON.stringify(part.value);
}

/** The `text` and `json` part texts of a message, joined without a separator. */
export function messageText(message: Message): string {
  let text = "";
  for (const part of message.content) {
    if (part.type === "text" || part.type === "json") text += partText(part);
  }
  return text;
}

/** True when `String.prototype.trim()` leaves nothing. */
export function isBlank(text: string): boolean {
  return text.trim() === "";
}

export function systemReminder(text: string): string {
  return `<system-reminder>\n${text}\n</system-reminder>`;
}

/**
 * Keeps a `tool.call` or `tool.result` part only when both the call and the result
 * with its `callId` exist in the list, then drops messages left without parts.
 */
export function repairToolPairs(messages: readonly Message[]): Message[] {
  const calls = new Set<string>();
  const results = new Set<string>();
  for (const message of messages) {
    for (const part of message.content) {
      if (part.type === "tool.call") calls.add(part.callId);
      else if (part.type === "tool.result") results.add(part.callId);
    }
  }
  const repaired: Message[] = [];
  for (const message of messages) {
    const content = message.content.filter((part) =>
      (part.type !== "tool.call" && part.type !== "tool.result") || (calls.has(part.callId) && results.has(part.callId)));
    if (content.length === 0) continue;
    repaired.push(content.length === message.content.length ? message : { ...message, content });
  }
  return repaired;
}

/** Splits off the consecutive `system` messages at the start; their non-blank texts follow the system blocks. */
export function splitLeadingSystem(messages: readonly Message[]): { instructions: string[]; rest: Message[] } {
  const instructions: string[] = [];
  let start = 0;
  for (; start < messages.length; start += 1) {
    const message = messages[start];
    if (message === undefined || message.role !== "system") break;
    const text = messageText(message);
    if (!isBlank(text)) instructions.push(text);
  }
  return { instructions, rest: messages.slice(start) };
}

/** Adds `type: object` to a tool input schema that has no `type` key. */
export function objectSchema(schema: Record<string, Json>): Record<string, Json> {
  return Object.hasOwn(schema, "type") ? schema : { type: "object", ...schema };
}

export type MediaSource =
  | { kind: "data"; data: string; mediaType: string }
  | { kind: "url"; url: string; mediaType: string };

function nonEmptyString(value: unknown): string | undefined {
  return typeof value === "string" && value !== "" ? value : undefined;
}

/** Resolves a `media` part with the host resolver. */
export async function resolveMediaPart(
  provider: ModelProvider,
  part: MediaPart,
  resolver: MediaResolver | undefined,
  signal: AbortSignal,
): Promise<MediaSource> {
  if (resolver === undefined) throw unsupportedContent(provider, `media part ${part.ref} without a resolveMedia function`);
  const resolved: unknown = await resolver({ ref: part.ref, mediaType: part.mediaType }, { signal });
  if (!isRecord(resolved)) throw invalidRequest(provider, `resolveMedia returned no data or url for ${part.ref}`);
  const mediaType = nonEmptyString(resolved.mediaType) ?? part.mediaType;
  if (typeof resolved.data === "string") return { kind: "data", data: resolved.data, mediaType };
  if (typeof resolved.url === "string") return { kind: "url", url: resolved.url, mediaType };
  throw invalidRequest(provider, `resolveMedia returned no data or url for ${part.ref}`);
}

export function isImageType(mediaType: string): boolean {
  return mediaType.startsWith("image/");
}

export const PDF_TYPE = "application/pdf";

File: python/goondan/goondan/models/_normalize.py
"""§대화 정규화: the provider-independent normalization both adapters run before mapping."""

from __future__ import annotations

import inspect
from typing import Any, Mapping, Sequence

from ._errors import invalid_request, unsupported_content
from ._options import MediaResolver
from ._values import is_blank, json_text, non_empty_text

PDF_TYPE = "application/pdf"


def part_text(part: Mapping[str, Any]) -> str:
    """The text of a `text` part, or the compact JSON text of a `json` part."""
    if part.get("type") == "text":
        text = part.get("text")
        return text if isinstance(text, str) else ""
    return json_text(part.get("value"))


def message_text(message: Mapping[str, Any]) -> str:
    """The `text` and `json` part texts of a message, joined without a separator."""
    text = ""
    for part in content_of(message):
        if part.get("type") in ("text", "json"):
            text += part_text(part)
    return text


def content_of(message: Mapping[str, Any]) -> list[dict[str, Any]]:
    content = message.get("content")
    if not isinstance(content, list):
        return []
    return [part for part in content if isinstance(part, dict)]


def system_reminder(text: str) -> str:
    return f"<system-reminder>\n{text}\n</system-reminder>"


def repair_tool_pairs(messages: Sequence[Mapping[str, Any]]) -> list[dict[str, Any]]:
    """§대화 정규화 1: keeps a tool part only when both the call and the result exist, then drops empty messages."""
    calls: set[str] = set()
    results: set[str] = set()
    for message in messages:
        for part in content_of(message):
            call_id = part.get("callId")
            if not isinstance(call_id, str):
                continue
            if part.get("type") == "tool.call":
                calls.add(call_id)
            elif part.get("type") == "tool.result":
                results.add(call_id)
    paired = calls & results
    repaired: list[dict[str, Any]] = []
    for message in messages:
        content = content_of(message)
        kept = [
            part for part in content
            if part.get("type") not in ("tool.call", "tool.result") or part.get("callId") in paired
        ]
        if not kept:
            continue
        repaired.append({**message, "content": kept})
    return repaired


def split_leading_system(messages: Sequence[Mapping[str, Any]]) -> tuple[list[str], list[dict[str, Any]]]:
    """§대화 정규화 2: splits off the consecutive `system` messages at the start."""
    instructions: list[str] = []
    start = 0
    while start < len(messages):
        message = messages[start]
        if message.get("role") != "system":
            break
        text = message_text(message)
        if not is_blank(text):
            instructions.append(text)
        start += 1
    return instructions, [dict(message) for message in messages[start:]]


def object_schema(schema: Any) -> dict[str, Any]:
    """§대화 정규화: a tool input schema without a `type` key is sent as an object schema."""
    if not isinstance(schema, dict):
        return {"type": "object"}
    return dict(schema) if "type" in schema else {"type": "object", **schema}


class MediaSource:
    """§미디어 해석: base64 data or a URL, with the media type the adapter must use."""

    __slots__ = ("kind", "data", "url", "media_type")

    def __init__(self, kind: str, media_type: str, *, data: str = "", url: str = ""):
        self.kind = kind
        self.media_type = media_type
        self.data = data
        self.url = url


async def resolve_media_part(provider: str, part: Mapping[str, Any], resolver: MediaResolver | None) -> MediaSource:
    """§미디어 해석: turns a `media` part into data or a URL with the host resolver."""
    ref = part.get("ref")
    reference = ref if isinstance(ref, str) else ""
    declared = part.get("mediaType")
    declared_type = declared if isinstance(declared, str) else ""
    if resolver is None:
        raise unsupported_content(provider, f"media part {reference} without a resolve_media function")
    resolved: Any = resolver({"ref": reference, "mediaType": declared_type})
    if inspect.isawaitable(resolved):
        resolved = await resolved
    if not isinstance(resolved, Mapping):
        raise invalid_request(provider, f"resolve_media returned no data or url for {reference}")
    media_type = non_empty_text(resolved.get("mediaType")) or declared_type
    data = resolved.get("data")
    if isinstance(data, str):
        return MediaSource("data", media_type, data=data)
    url = resolved.get("url")
    if isinstance(url, str):
        return MediaSource("url", media_type, url=url)
    raise invalid_request(provider, f"resolve_media returned no data or url for {reference}")


def is_image_type(media_type: str) -> bool:
    return media_type.startswith("image/")
